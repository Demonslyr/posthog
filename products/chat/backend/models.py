from typing import TYPE_CHECKING

from django.db import models
from django.db.models import QuerySet

from posthog.models.utils import UUIDModel
from posthog.models.file_system.file_system_mixin import FileSystemSyncMixin
from posthog.models.file_system.file_system_representation import FileSystemRepresentation

if TYPE_CHECKING:
    from posthog.models.team import Team


class ChatConversation(FileSystemSyncMixin, UUIDModel):
    class Meta:
        db_table = "posthog_chat_conversation"
        # No need for the unique constraint since we removed conversation_id
        pass

    person = models.ForeignKey(
        "posthog.Person",
        on_delete=models.CASCADE,
        related_name="chat_conversations",
        related_query_name="chat_conversation",
    )

    team = models.ForeignKey(
        "posthog.Team",
        on_delete=models.CASCADE,
        related_name="all_chat_conversations",
        related_query_name="all_chat_conversation",
    )

    title = models.CharField(max_length=400, blank=True, null=True)

    # Metadata
    created_at = models.DateTimeField(auto_now_add=True)
    updated_at = models.DateTimeField(auto_now=True)

    # Which page/url the conversation started on
    source_url = models.URLField(max_length=2000, blank=True, null=True)

    # Unread message count
    unread_count = models.PositiveIntegerField(default=0)

    @classmethod
    def get_file_system_unfiled(cls, team: "Team") -> QuerySet["ChatConversation"]:
        base_qs = cls.objects.filter(team=team)
        return cls._filter_unfiled_queryset(base_qs, team, type="chat_conversation", ref_field="id")

    def get_file_system_representation(self) -> FileSystemRepresentation:
        return FileSystemRepresentation(
            base_folder=self._create_in_folder or "Unfiled/Chats",
            type="chat_conversation",
            ref=str(self.pk),
            name=self.title
            or f"Chat with {self.person.properties.get('name') or self.person.properties.get('email') or f'Person {self.person.id}'}",
            href=f"/chat/{self.pk}",
            meta={
                "created_at": str(self.created_at),
                "person_id": str(self.person.id),
            },
            should_delete=False,
        )


class ChatMessage(UUIDModel):
    conversation = models.ForeignKey(
        ChatConversation,
        on_delete=models.CASCADE,
        related_name="messages",
        related_query_name="message",
    )

    content = models.TextField()

    # Metadata
    created_at = models.DateTimeField(auto_now_add=True)

    # Message has been read by the other party
    read = models.BooleanField(default=False)

    is_assistant = models.BooleanField(default=False, db_column="is_assistant")

    class Meta:
        ordering = ["created_at"]
        db_table = "posthog_chat_message"
