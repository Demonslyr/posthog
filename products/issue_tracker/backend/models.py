from django.db import models
from django.utils import timezone
import uuid


class Issue(models.Model):
    class Status(models.TextChoices):
        BACKLOG = "backlog", "Backlog"
        TODO = "todo", "To Do"
        IN_PROGRESS = "in_progress", "In Progress"
        TESTING = "testing", "Testing"
        DONE = "done", "Done"

    class OriginProduct(models.TextChoices):
        ERROR_TRACKING = "error_tracking", "Error Tracking"
        EVAL_CLUSTERS = "eval_clusters", "Eval Clusters"
        USER_CREATED = "user_created", "User Created"
        SUPPORT_QUEUE = "support_queue", "Support Queue"

    id = models.UUIDField(primary_key=True, default=uuid.uuid4, editable=False)
    team = models.ForeignKey("posthog.Team", on_delete=models.CASCADE)
    title = models.CharField(max_length=255)
    description = models.TextField()
    status = models.CharField(max_length=20, choices=Status.choices, default=Status.BACKLOG)
    origin_product = models.CharField(max_length=20, choices=OriginProduct.choices)
    position = models.IntegerField(default=0)
    created_at = models.DateTimeField(default=timezone.now)
    updated_at = models.DateTimeField(auto_now=True)

    class Meta:
        db_table = "posthog_issue"
        managed = True
        ordering = ["position"]

    def __str__(self):
        return f"{self.title} ({self.get_status_display()})"
