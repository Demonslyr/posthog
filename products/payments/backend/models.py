from typing import TYPE_CHECKING
from django.db import models
from posthog.models.utils import UUIDModel, sane_repr

if TYPE_CHECKING:
    pass


class PaymentProduct(UUIDModel):
    class Meta:
        db_table = "posthog_payments_products"

    team = models.ForeignKey("posthog.Team", on_delete=models.CASCADE)
    stripe_product_id = models.CharField(max_length=100)
    stripe_pricing_id = models.CharField(max_length=100)
    name = models.CharField(max_length=200)
    description = models.TextField(blank=True)
    price = models.DecimalField(max_digits=10, decimal_places=2)
    currency = models.CharField(max_length=3)  # e.g. USD, EUR
    date_created = models.DateTimeField(auto_now_add=True)
    date_updated = models.DateTimeField(auto_now=True)
    deleted = models.BooleanField(default=False)
    is_active = models.BooleanField(default=True)

    def __str__(self) -> str:
        return self.name

    __repr__ = sane_repr("id", "name", "team_id", "stripe_product_id")


class PaymentTransaction(UUIDModel):
    class Meta:
        db_table = "posthog_payments_transactions"

    team = models.ForeignKey("posthog.Team", on_delete=models.CASCADE)
    product = models.ForeignKey(PaymentProduct, on_delete=models.PROTECT)
    payload = models.JSONField()
    status = models.CharField(max_length=50)
    date_created = models.DateTimeField(auto_now_add=True)

    def __str__(self) -> str:
        return f"{self.team_id} - {self.product.name} - {self.status}"

    __repr__ = sane_repr("id", "team_id", "product_id", "status")
