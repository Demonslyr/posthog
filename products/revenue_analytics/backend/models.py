from typing import cast, Any, Optional
import posthoganalytics

from posthog.hogql import ast
from posthog.warehouse.models.external_data_source import ExternalDataSource
from posthog.warehouse.models.external_data_schema import ExternalDataSchema
from posthog.warehouse.models.table import DataWarehouseTable
from posthog.models.exchange_rate.sql import EXCHANGE_RATE_DECIMAL_PRECISION
from posthog.schema import CurrencyCode
from posthog.hogql.database.models import (
    BooleanDatabaseField,
    DateTimeDatabaseField,
    ExpressionField,
    IntegerDatabaseField,
    StringDatabaseField,
    Table,
    FieldOrTable,
)
from posthog.hogql.database.schema.exchange_rate import (
    DEFAULT_CURRENCY,
    convert_currency_call,
)


from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from posthog.hogql.context import HogQLContext

STRIPE_DATA_WAREHOUSE_CHARGE_IDENTIFIER = "Charge"

# Stripe represents most currencies with integer amounts multiplied by 100,
# since most currencies have its smallest unit as 1/100 of their base unit
# It just so happens that some currencies don't have that concept, so they're listed here
# https://docs.stripe.com/currencies#zero-decimal
ZERO_DECIMAL_CURRENCIES_IN_STRIPE: list[CurrencyCode] = [
    CurrencyCode.BIF,
    CurrencyCode.CLP,
    CurrencyCode.DJF,
    CurrencyCode.GNF,
    CurrencyCode.JPY,
    CurrencyCode.KMF,
    CurrencyCode.KRW,
    CurrencyCode.MGA,
    CurrencyCode.PYG,
    CurrencyCode.RWF,
    CurrencyCode.UGX,
    CurrencyCode.VND,
    CurrencyCode.VUV,
    CurrencyCode.XAF,
    CurrencyCode.XOF,
    CurrencyCode.XPF,
]

BASE_FIELDS: dict[str, FieldOrTable] = {
    "distinct_id": StringDatabaseField(name="id"),
    "__original_amount": IntegerDatabaseField(name="amount", hidden=True),
    "original_amount": ExpressionField(
        isolate_scope=True,
        expr=ast.Call(
            name="toDecimal",
            args=[
                ast.Field(chain=["__original_amount"]),
                ast.Constant(value=EXCHANGE_RATE_DECIMAL_PRECISION),
            ],
        ),
        name="original_amount",
    ),
    "is_refund": BooleanDatabaseField(name="refunded"),
    "__original_currency": StringDatabaseField(name="currency", hidden=True),
    "original_currency": ExpressionField(
        isolate_scope=True,
        expr=ast.Call(name="upper", args=[ast.Field(chain=["__original_currency"])]),
        name="original_currency",
    ),
    # As per above, Stripe represents most currencies with integer amounts multiplied by 100,
    # so we need to divide by 100 to get the actual amount
    "currency_is_zero_decimal": ExpressionField(
        isolate_scope=True,
        expr=ast.CompareOperation(
            op=ast.CompareOperationOp.In,
            left=ast.Field(chain=["original_currency"]),
            right=ast.Array(
                exprs=[ast.Constant(value=currency.value) for currency in ZERO_DECIMAL_CURRENCIES_IN_STRIPE]
            ),
        ),
        name="currency_is_zero_decimal",
        hidden=True,
    ),
    "amount_decimal_divider": ExpressionField(
        isolate_scope=True,
        expr=ast.Call(
            name="if",
            args=[
                ast.Field(chain=["currency_is_zero_decimal"]),
                ast.Call(
                    name="toDecimal", args=[ast.Constant(value=1), ast.Constant(value=EXCHANGE_RATE_DECIMAL_PRECISION)]
                ),
                ast.Call(
                    name="toDecimal",
                    args=[ast.Constant(value=100), ast.Constant(value=EXCHANGE_RATE_DECIMAL_PRECISION)],
                ),
            ],
        ),
        name="amount_decimal_divider",
        hidden=True,
    ),
    # Amount taking into consideration whether this was a refund or not
    # plus the decimal divider
    "adjusted_original_amount": ExpressionField(
        isolate_scope=True,
        expr=ast.Call(
            name="divideDecimal",
            args=[
                ast.Call(
                    name="if",
                    args=[
                        ast.Field(chain=["is_refund"]),
                        ast.Call(name="negate", args=[ast.Field(chain=["original_amount"])]),
                        ast.Field(chain=["original_amount"]),
                    ],
                ),
                ast.Field(chain=["amount_decimal_divider"]),
            ],
        ),
        name="adjusted_original_amount",
    ),
    # Allow us to sort by timestamp properly in insights (and split in buckets)
    "timestamp": DateTimeDatabaseField(name="created_at"),
}


class RevenueAnalyticsRevenueView(Table):
    data_warehouse_table: (
        Any  # Pydantic is complaining for reasons I dont understand, this should be DataWarehouseTable
    )

    @staticmethod
    def for_schema_source(source: ExternalDataSource) -> Optional["RevenueAnalyticsRevenueView"]:
        # Currently only works for stripe sources
        if not source.source_type == ExternalDataSource.Type.STRIPE:
            return None

        # The table we care about is the one with schema `Charge` since from there we can get
        # the data we need in our view
        try:
            schema: ExternalDataSchema = ExternalDataSchema.objects.get(
                source=source, name=STRIPE_DATA_WAREHOUSE_CHARGE_IDENTIFIER
            )
            table: Optional[DataWarehouseTable] = cast(
                Optional[DataWarehouseTable], schema.table
            )  # Weird cast because pydantic is weird
        except (ExternalDataSchema.DoesNotExist, DataWarehouseTable.DoesNotExist):
            return None

        if table is None:
            return None

        team = table.team
        revenue_config = team.revenue_config
        do_currency_conversion = posthoganalytics.feature_enabled(
            "web-analytics-revenue-tracking-conversion",
            str(team.organization_id),
            groups={"organization": str(team.organization_id)},
            group_properties={"organization": {"id": str(team.organization_id)}},
        )

        # Get most of the fields directly from the base fields,
        # but generate the converted currency and amounts dynamically
        # since we depend on the account's configuration
        fields: dict[str, FieldOrTable] = {
            **BASE_FIELDS,
            "currency": ExpressionField(
                isolate_scope=True,
                expr=ast.Constant(value=(revenue_config.baseCurrency or DEFAULT_CURRENCY).value),
                name="currency",
            )
            if do_currency_conversion
            else BASE_FIELDS["original_currency"],
            "amount": ExpressionField(
                isolate_scope=True,
                expr=convert_currency_call(
                    # Only doing the adjusted calculation here since this is after we've divided by 100
                    ast.Field(chain=["adjusted_original_amount"]),
                    ast.Field(chain=["original_currency"]),
                    ast.Field(chain=["currency"]),
                    ast.Call(
                        name="_toDate",
                        args=[
                            # Because timestamp columns are nullable, we need to handle that case
                            # by converting to a default value of 0
                            ast.Call(
                                name="ifNull",
                                args=[
                                    ast.Field(chain=["timestamp"]),
                                    ast.Call(name="toDateTime", args=[ast.Constant(value=0)]),
                                ],
                            )
                        ],
                    ),
                ),
                name="amount",
            )
            if do_currency_conversion
            else BASE_FIELDS["adjusted_original_amount"],
            # Required because we add a "team_id" where clause on every selected table
            "team_id": ExpressionField(
                isolate_scope=True,
                expr=ast.Constant(value=team.pk),
                name="team_id",
            ),
        }

        return RevenueAnalyticsRevenueView(fields=fields, data_warehouse_table=table)

    def to_printed_clickhouse(self, context: "HogQLContext") -> str:
        table = cast(DataWarehouseTable, self.data_warehouse_table)
        return table.name

    def to_printed_hogql(self) -> str:
        table = cast(DataWarehouseTable, self.data_warehouse_table)
        return table.name
