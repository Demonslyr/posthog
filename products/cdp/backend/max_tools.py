from typing import Optional
from common.hogvm.python.execute import validate_bytecode
from ee.hogai.tool import MaxTool
from posthog.cdp.validation import compile_hog
from posthog.hogql.ai import HOG_EXAMPLE_MESSAGE, HOG_GRAMMAR_MESSAGE, IDENTITY_MESSAGE_HOG
from products.cdp.backend.prompts import HOG_TRANSFORMATION_ASSISTANT_ROOT_SYSTEM_PROMPT
from pydantic import BaseModel, Field
from langchain_openai import ChatOpenAI
from langchain_core.prompts import ChatPromptTemplate
from ee.hogai.graph.schema_generator.parsers import PydanticOutputParserException, parse_pydantic_structured_output


class CreateHogTransformationFunctionArgs(BaseModel):
    instructions: str = Field(description="The instructions for what transformation to create.")


class HogTransformationOutput(BaseModel):
    hog_code: str


class CreateHogTransformationFunctionTool(MaxTool):
    name: str = "create_hog_transformation_function"  # Must match a value in AssistantContextualTool enum
    description: str = (
        "Write or edit the hog code to create your desired transformation and apply it to the current editor"
    )
    thinking_message: str = "Creating your desired transformation"
    args_schema: type[BaseModel] = CreateHogTransformationFunctionArgs
    root_system_prompt_template: str = HOG_TRANSFORMATION_ASSISTANT_ROOT_SYSTEM_PROMPT

    def _run_impl(self, instructions: str) -> tuple[str, str]:
        prompt = ChatPromptTemplate.from_messages(
            [
                (
                    "system",
                    IDENTITY_MESSAGE_HOG
                    + "\n\n<example_hog_code>\n"
                    + HOG_EXAMPLE_MESSAGE
                    + "\n</example_hog_code>\n\n"
                    + "\n\n<hog_grammar>\n"
                    + HOG_GRAMMAR_MESSAGE
                    + "\n</hog_grammar>\n\n"
                    + "\n\n<current_hog_code>\n{{{current_hog_code}}}\n</current_hog_code>"
                    + "\n\nRemove all line breaks and carriage returns, and strip whitespace."
                    + "\n\nReturn the hog code nicely formatted.",
                ),
                (
                    "user",
                    "Write a Hog transformation or tweak the current one to satisfy this request: " + instructions,
                ),
            ],
            template_format="mustache",
        )

        final_error: Optional[Exception] = None
        for _ in range(3):
            try:
                chain = prompt | self._model
                result = chain.invoke(self.context)
                parsed_result = self._parse_output(result)
                break
            except PydanticOutputParserException as e:
                prompt += f"Avoid this error: {str(e)}"
                final_error = e
        else:
            raise final_error

        return "```hog\n" + parsed_result.hog_code + "\n```", parsed_result.hog_code

    @property
    def _model(self):
        return ChatOpenAI(model="gpt-4o", temperature=0, disable_streaming=True).with_structured_output(
            HogTransformationOutput,
            method="function_calling",
            include_raw=False,
        )

    def _parse_output(self, output):  # type: ignore
        result = parse_pydantic_structured_output(HogTransformationOutput)(output)
        assert result is not None

        # Validate that the Hog code compiles to bytecode
        try:
            compiled_result = compile_hog(result.hog_code, "transformation")
            if compiled_result:
                is_valid, error_message = validate_bytecode(compiled_result, {})
                if not is_valid:
                    raise PydanticOutputParserException(
                        llm_output=result.hog_code,
                        validation_message=f"The Hog bytecode validation failed: {error_message}",
                    )
        except Exception as e:
            raise PydanticOutputParserException(
                llm_output=result.hog_code, validation_message=f"The Hog code failed to compile: {str(e)}"
            )

        return result
