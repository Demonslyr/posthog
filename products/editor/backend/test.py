import json
from datetime import datetime

from pydantic import BaseModel, Field

from posthog.clickhouse.client.execute import sync_execute
from posthog.test.base import BaseTest
from products.editor.backend.models.codebase import Codebase
from products.editor.backend.services.codebase_sync import SerializedArtifact


class CatalogEntry(BaseModel):
    team_id: int | None = Field(default=None)
    user_id: int | None = Field(default=None)
    codebase_id: str | None = Field(default=None)
    artifact_id: str
    parent_artifact_id: str | None = Field(default=None)
    branch: str | None = Field(default=None)
    type: str = Field(default="dir")
    sign: int = Field(default=1)
    timestamp: datetime | None = Field(default=None)


class EditorTestQueryHelpersMixin(BaseTest):
    codebase: Codebase
    stable_user_id: int = 99999

    def _create_artifacts(self, tree: list[SerializedArtifact]):
        query = "INSERT INTO codebase_embeddings (team_id, user_id, codebase_id, artifact_id, chunk_id, vector, properties, is_deleted) VALUES "
        rows: list[str] = []

        args = {
            "team_id": self.team.id,
            "user_id": self.user.id,
            "codebase_id": self.codebase.id,
            "chunk_id": 0,
            "vector": [0.5, 0.5],
            "properties": json.dumps(
                {
                    "lineStart": 0,
                    "lineEnd": 30,
                    "path": "obfuscated_path",
                }
            ),
            "is_deleted": 0,
        }

        for idx, node in enumerate(tree):
            if node["type"] == "file":
                args.update(
                    {
                        f"artifact_id_{idx}": node["id"],
                    }
                )
                rows.append(
                    f"(%(team_id)s, %(user_id)s, %(codebase_id)s, %(artifact_id_{idx})s, %(chunk_id)s, %(vector)s, %(properties)s, %(is_deleted)s)"
                )

        sync_execute(query + ", ".join(rows), args, team_id=self.team.id)

    def _create_codebase_catalog(self, tree: list[CatalogEntry]):
        query = "INSERT INTO codebase_catalog (team_id, user_id, codebase_id, artifact_id, parent_artifact_id, branch, type, timestamp, sign) VALUES "
        rows: list[str] = []
        args = {}

        for idx, node in enumerate(tree):
            args.update(
                {
                    f"team_id_{idx}": node.team_id or self.team.id,
                    f"user_id_{idx}": node.user_id or self.stable_user_id,
                    f"codebase_id_{idx}": node.codebase_id or str(self.codebase.id),
                    f"artifact_id_{idx}": node.artifact_id,
                    f"parent_artifact_id_{idx}": node.parent_artifact_id,
                    f"branch_{idx}": node.branch or "",
                    f"type_{idx}": node.type,
                    f"timestamp_{idx}": node.timestamp,
                    f"sign_{idx}": node.sign,
                }
            )
            rows.append(
                f"(%(team_id_{idx})s, %(user_id_{idx})s, %(codebase_id_{idx})s, %(artifact_id_{idx})s, %(parent_artifact_id_{idx})s, %(branch_{idx})s, %(type_{idx})s, %(timestamp_{idx})s, %(sign_{idx})s)"
            )

        sync_execute(query + ", ".join(rows), args, team_id=self.team.id)
