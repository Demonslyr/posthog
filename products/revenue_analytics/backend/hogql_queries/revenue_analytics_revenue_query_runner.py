from posthog.hogql import ast
from posthog.hogql.query import execute_hogql_query
from posthog.schema import (
    CachedRevenueAnalyticsRevenueQueryResponse,
    RevenueAnalyticsRevenueQueryResponse,
    RevenueAnalyticsRevenueQuery,
    QueryDateRangeResponse,
)

from .revenue_analytics_query_runner import RevenueAnalyticsQueryRunner


class RevenueAnalyticsRevenueQueryRunner(RevenueAnalyticsQueryRunner):
    query: RevenueAnalyticsRevenueQuery
    response: RevenueAnalyticsRevenueQueryResponse
    cached_response: CachedRevenueAnalyticsRevenueQueryResponse

    def to_query(self) -> ast.SelectQuery:
        return ast.SelectQuery.empty()

    def calculate(self):
        response = execute_hogql_query(
            query_type="revenue_analytics_revenue_query",
            query=self.to_query(),
            team=self.team,
            timings=self.timings,
            modifiers=self.modifiers,
            limit_context=self.limit_context,
        )

        return RevenueAnalyticsRevenueQueryResponse(
            results=response.results,
            hogql=response.hogql,
            modifiers=self.modifiers,
            query_date_range=QueryDateRangeResponse(
                date_from=self.query_date_range.date_from(),
                date_to=self.query_date_range.date_to(),
                interval=self.query_date_range.interval_type,
            ),
        )
