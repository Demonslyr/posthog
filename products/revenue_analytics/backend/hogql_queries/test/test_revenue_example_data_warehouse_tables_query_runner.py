from freezegun import freeze_time
from unittest.case import skip
from unittest.mock import patch

from products.revenue_analytics.backend.hogql_queries.revenue_example_data_warehouse_tables_query_runner import (
    RevenueExampleDataWarehouseTablesQueryRunner,
)
from products.revenue_analytics.backend.models import STRIPE_DATA_WAREHOUSE_CHARGE_IDENTIFIER

from posthog.schema import (
    RevenueExampleDataWarehouseTablesQuery,
    RevenueTrackingConfig,
    RevenueExampleDataWarehouseTablesQueryResponse,
    CurrencyCode,
)
from posthog.test.base import (
    APIBaseTest,
    ClickhouseTestMixin,
    snapshot_clickhouse_queries,
)
from posthog.warehouse.models import (
    DataWarehouseTable,
    DataWarehouseCredential,
    ExternalDataSource,
    ExternalDataSchema,
)

REVENUE_TRACKING_CONFIG = RevenueTrackingConfig(baseCurrency=CurrencyCode.GBP, events=[])


@snapshot_clickhouse_queries
class TestRevenueExampleDataWarehouseTablesQueryRunner(ClickhouseTestMixin, APIBaseTest):
    QUERY_TIMESTAMP = "2025-01-29"

    def setUp(self):
        super().setUp()

        self.source = ExternalDataSource.objects.create(
            team=self.team,
            source_id="source_id",
            connection_id="connection_id",
            status=ExternalDataSource.Status.COMPLETED,
            source_type=ExternalDataSource.Type.STRIPE,
        )

        self.credential = DataWarehouseCredential.objects.create(
            team=self.team,
            access_key="test",
            access_secret="test",
        )

        self.table = DataWarehouseTable.objects.create(
            name="stripe_charge",
            format=DataWarehouseTable.TableFormat.Parquet,
            team=self.team,
            credential=self.credential,
            url_pattern="test://localhost",
            columns={
                "id": {"hogql": "StringDatabaseField", "clickhouse": "String", "schema_valid": True},
                "revenue": {"hogql": "FloatDatabaseField", "clickhouse": "Float64", "schema_valid": True},
                "timestamp": {"hogql": "DateTimeDatabaseField", "clickhouse": "DateTime", "schema_valid": True},
            },
        )

        self.schema = ExternalDataSchema.objects.create(
            team=self.team,
            name=STRIPE_DATA_WAREHOUSE_CHARGE_IDENTIFIER,
            source=self.source,
            table=self.table,
            should_sync=True,
            last_synced_at="2024-01-01",
        )

    def _run_revenue_example_external_tables_query(self):
        with freeze_time(self.QUERY_TIMESTAMP):
            query = RevenueExampleDataWarehouseTablesQuery(revenueTrackingConfig=REVENUE_TRACKING_CONFIG)
            runner = RevenueExampleDataWarehouseTablesQueryRunner(team=self.team, query=query)

            response = runner.calculate()
            RevenueExampleDataWarehouseTablesQueryResponse.model_validate(response)

            return response

    def test_no_crash_when_no_data(self):
        self.table.delete()
        results = self._run_revenue_example_external_tables_query().results

        assert len(results) == 0

    # TODO: These fail because it'll complain the table doesn't exist
    # posthog.errors.CHQueryErrorUnknownTable: Code: 60.
    # DB::Exception: Unknown table expression identifier 'stripe_charge'
    @skip("Skipping because it's not implemented in tests")
    @patch("posthoganalytics.feature_enabled", return_value=False)
    def test_database_query(self, feature_enabled_mock):
        response = self._run_revenue_example_external_tables_query()
        results = response.results

        assert len(results) == 3

    # TODO: These fail because it'll complain the table doesn't exist
    # posthog.errors.CHQueryErrorUnknownTable: Code: 60.
    # DB::Exception: Unknown table expression identifier 'stripe_charge'
    @skip("Skipping because it's not implemented in tests")
    @patch("posthoganalytics.feature_enabled", return_value=True)
    def test_database_query_with_currency_conversion(self, feature_enabled_mock):
        response = self._run_revenue_example_external_tables_query()
        results = response.results

        assert len(results) == 3
