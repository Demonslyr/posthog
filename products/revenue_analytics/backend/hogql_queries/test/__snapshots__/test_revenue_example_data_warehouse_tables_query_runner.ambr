# serializer version: 1
# name: TestRevenueExampleDataWarehouseTablesQueryRunner.test_no_crash_when_no_data
  '''
  SELECT 1
  WHERE 0
  LIMIT 101
  OFFSET 0 SETTINGS readonly=2,
                    max_execution_time=60,
                    allow_experimental_object_type=1,
                    format_csv_allow_double_quotes=0,
                    max_ast_elements=4000000,
                    max_expanded_ast_elements=4000000,
                    max_bytes_before_external_group_by=0,
                    transform_null_in=1
  '''
# ---
